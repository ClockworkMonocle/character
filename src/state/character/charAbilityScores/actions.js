import invariant from 'invariant';

export const editAbilityScores = scores => dispatch => {
  invariant(
    typeof scores.str === 'number',
    'Ability score `str` must be number'
  );
  invariant(
    typeof scores.dex === 'number',
    'Ability score `dex` must be number'
  );
  invariant(
    typeof scores.con === 'number',
    'Ability score `con` must be number'
  );
  invariant(
    typeof scores.wis === 'number',
    'Ability score `wis` must be number'
  );
  invariant(
    typeof scores.int === 'number',
    'Ability score `int` must be number'
  );
  invariant(
    typeof scores.cha === 'number',
    'Ability score `cha` must be number'
  );

  dispatch({
    type: 'ABILITY_SCORE_EDIT',
    data: scores,
  });
};
