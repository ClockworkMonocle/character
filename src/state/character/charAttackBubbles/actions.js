import invariant from 'invariant';

const checkObject = data => {
  invariant(
    typeof data.abil === 'string',
    'Attack bonus `abil` must be a string'
  );
  invariant(
    typeof data.prof === 'boolean',
    'Attack bonus `prof` must be a boolean'
  );
  invariant(
    typeof data.name === 'string',
    'Attack bonus `name` must be a string'
  );
  invariant(typeof data.id === 'string', 'Attack bonus `id` must be a string');
  invariant(
    typeof data.bonus === 'number',
    'Attack bonus `bonus` must be a number'
  );

  return true;
};

export const createAttackBubble = data => dispatch =>
  checkObject(data) &&
  dispatch({
    type: 'ATTACK_BONUS_CREATE',
    data,
  });

export const editAttackBubble = data => dispatch =>
  checkObject(data) &&
  dispatch({
    type: 'ATTACK_BONUS_EDIT',
    data,
  });

export const deleteAttackBubble = data => dispatch => {
  invariant(typeof data.id === 'string', 'Attack bonus `id` must be a string');

  dispatch({
    type: 'ATTACK_BONUS_DELETE',
    data,
  });
};
