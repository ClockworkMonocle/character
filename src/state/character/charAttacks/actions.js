import invariant from 'invariant';

export const createAttack = data => dispatch => {
  invariant(typeof data.id === 'string', 'Create Attack `id` must be a string');

  invariant(
    typeof data.name === 'string',
    'Create Attack `name` must be a string'
  );

  invariant(
    typeof data.desc === 'string',
    'Create Attack `desc` must be a string'
  );

  dispatch({
    type: 'ATTACK_CREATE',
    data,
  });
};

export const editAttack = data => dispatch => {
  invariant(typeof data.id === 'string', 'Edit Attack `id` must be a string');

  invariant(
    typeof data.name === 'string',
    'Edit Attack `name` must be a string'
  );

  invariant(
    typeof data.desc === 'string',
    'Edit Attack `desc` must be a string'
  );

  dispatch({
    type: 'ATTACK_EDIT',
    data,
  });
};

export const deleteAttack = data => dispatch => {
  invariant(typeof data.id === 'string', 'Delete Attack `id` must be a string');

  dispatch({
    type: 'ATTACK_DELETE',
    data,
  });
};
