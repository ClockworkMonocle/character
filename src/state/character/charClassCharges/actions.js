import invariant from 'invariant';

export const incrementClassCharge = id => dispatch => {
  invariant(
    typeof id === 'string',
    'Increment class charge `id` must be a string'
  );

  dispatch({
    type: 'CLASS_CHARGE_INCREMENT',
    data: {
      id,
    },
  });
};

export const decrementClassCharge = id => dispatch => {
  invariant(
    typeof id === 'string',
    'Decrement class charge `id` must be a string'
  );

  dispatch({
    type: 'CLASS_CHARGE_DECREMENT',
    data: {
      id,
    },
  });
};
