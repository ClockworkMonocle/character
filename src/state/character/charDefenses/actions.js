import invariant from 'invariant';

export const editDefenses = data => dispatch => {
  invariant(
    typeof data.armorClass === 'number',
    'Edit defenses `armorClass` must be a number'
  );

  invariant(
    typeof data.initiativeBonus === 'number',
    'Edit defenses `initiativeBonus` must be a number'
  );

  invariant(
    typeof data.speed === 'string',
    'Edit defenses `speed` must be a string'
  );

  invariant(
    typeof data.maxHp === 'number',
    'Edit defenses `maxHp` must be a number'
  );

  dispatch({
    type: 'DEFENSES_EDIT',
    data,
  });
};
