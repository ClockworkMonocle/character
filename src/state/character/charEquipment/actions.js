import invariant from 'invariant';

export const addWealth = data => dispatch => {
  invariant(
    typeof data.wealthType === 'string',
    'Add wealth `wealthType` must be a string'
  );
  invariant(
    typeof data.value === 'number',
    'Add wealth `value` must be a number'
  );

  dispatch({
    type: 'WEALTH_EDIT',
    data: {
      ...data,
      actionType: 'add',
    },
  });
};

export const subtractWealth = data => dispatch => {
  invariant(
    typeof data.wealthType === 'string',
    'Subtract wealth `wealthType` must be a string'
  );
  invariant(
    typeof data.value === 'number',
    'Subtract wealth `value` must be a number'
  );

  dispatch({
    type: 'WEALTH_EDIT',
    data: {
      ...data,
      actionType: 'subtract',
    },
  });
};

export const createContainer = data => dispatch => {
  invariant(
    typeof data.id === 'string',
    'Create container requires `id` to be a string'
  );

  invariant(
    typeof data.name === 'string',
    'Create container requires `name` to be a string'
  );

  dispatch({
    type: 'EQUIPMENT_CONTAINER_CREATE',
    data,
  });
};

export const editContainer = data => dispatch => {
  invariant(
    typeof data.id === 'string',
    'Edit container requires `id` to be a string'
  );

  invariant(
    typeof data.name === 'string',
    'Edit container requires `name` to be a string'
  );

  dispatch({
    type: 'EQUIPMENT_CONTAINER_EDIT',
    data,
  });
};

export const deleteContainer = data => dispatch => {
  invariant(
    typeof data.id === 'string',
    'Delete container requires `id` to be a string'
  );

  dispatch({
    type: 'EQUIPMENT_CONTAINER_DELETE',
    data,
  });
};

export const createItem = (container, item) => dispatch => {
  invariant(
    typeof container.id === 'string',
    'Create items requires `id` of container to be a string'
  );

  invariant(
    typeof item.id === 'string',
    'Create items requires `id` of item to be a string'
  );

  invariant(
    typeof item.name === 'string',
    'Create items requires `name` of item to be a string'
  );

  invariant(
    typeof item.desc === 'string',
    'Create items requires `desc` of item to be a string'
  );

  dispatch({
    type: 'EQUIPMENT_ITEM_CREATE',
    data: {
      item,
      container,
    },
  });
};

export const editItem = (container, item) => dispatch => {
  invariant(
    typeof container.id === 'string',
    'Create items requires `id` of container to be a string'
  );

  invariant(
    typeof item.id === 'string',
    'Create items requires `id` of item to be a string'
  );

  invariant(
    typeof item.name === 'string',
    'Create items requires `name` of item to be a string'
  );

  invariant(
    typeof item.desc === 'string',
    'Create items requires `desc` of item to be a string'
  );

  dispatch({
    type: 'EQUIPMENT_ITEM_EDIT',
    data: {
      item,
      container,
    },
  });
};

export const moveItem = (fromContainer, container, item) => dispatch => {
  invariant(
    typeof fromContainer.id === 'string',
    'Create items requires `id` of source container to be a string'
  );

  invariant(
    typeof container.id === 'string',
    'Create items requires `id` of destination container to be a string'
  );

  invariant(
    typeof item.id === 'string',
    'Create items requires `id` of item to be a string'
  );

  dispatch({
    type: 'EQUIPMENT_ITEM_MOVE',
    data: {
      item,
      container,
      fromContainer,
    },
  });
};

export const deleteItem = (container, item) => dispatch => {
  invariant(
    typeof container.id === 'string',
    'Create items requires `id` of container to be a string'
  );

  invariant(
    typeof item.id === 'string',
    'Create items requires `id` of item to be a string'
  );

  dispatch({
    type: 'EQUIPMENT_ITEM_DELETE',
    data: {
      item,
      container,
    },
  });
};
