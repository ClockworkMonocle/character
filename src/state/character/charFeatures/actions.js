import invariant from 'invariant';

export const createFeature = (feature, classCharge) => dispatch => {
  if (classCharge) {
    invariant(
      typeof classCharge.featureId === 'string',
      'Creating a feature with class charge requires `featureId` to be a string'
    );

    invariant(
      classCharge.featureId === feature.id,
      'Class charge `featureId` must equal the feauture id'
    );

    invariant(
      feature.classChargesId === classCharge.id,
      "A feature's `classChargesId` must match the class charge `id`"
    );
  }

  dispatch({
    type: 'FEATURE_CREATE',
    data: {
      feature,
      classCharge,
    },
  });
};

export const editFeature = (feature, classCharge) => dispatch => {
  if (classCharge) {
    invariant(
      typeof classCharge.featureId === 'string',
      'Creating a feature with class charge requires `featureId` to be a string'
    );

    invariant(
      classCharge.featureId === feature.id,
      'Class charge `featureId` must equal the feauture id'
    );

    invariant(
      feature.classChargesId === classCharge.id,
      "A feature's `classChargesId` must match the class charge `id`"
    );
  }

  dispatch({
    type: 'FEATURE_EDIT',
    data: {
      feature,
      classCharge,
    },
  });
};

export const deleteFeature = feature => dispatch =>
  dispatch({
    type: 'FEATURE_DELETE',
    data: {
      feature,
    },
  });
