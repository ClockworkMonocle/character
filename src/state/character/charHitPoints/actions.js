import invariant from 'invariant';

export const healCharacter = value => dispatch => {
  invariant(
    typeof value === 'number',
    'Heal character `value` must be a number'
  );

  dispatch({
    type: 'HIT_POINTS_HEAL',
    data: {
      value,
    },
  });
};

export const damageCharacter = value => dispatch => {
  invariant(
    typeof value === 'number',
    'Damage character `value` must be a number'
  );

  dispatch({
    type: 'HIT_POINTS_DAMAGE',
    data: {
      value,
    },
  });
};

export const tempHpCharacter = value => dispatch => {
  invariant(
    typeof value === 'number',
    'TempHp character `value` must be a number'
  );

  dispatch({
    type: 'HIT_POINTS_TEMP',
    data: {
      value,
    },
  });
};

export const deathSaveSuccessAdd = () => ({
  type: 'DEATH_SAVES_SUCCESS_ADD',
});

export const deathSaveFailureAdd = () => ({
  type: 'DEATH_SAVES_FAILURE_ADD',
});

export const createHitDice = () => ({
  type: 'HIT_DICE_CREATE',
});

export const editHitDice = data => dispatch => {
  invariant(typeof data.id === 'string', 'Edit hit dice `id` must be a string');

  invariant(
    typeof data.type === 'string',
    'Edit hit dice `type` must be a string'
  );

  invariant(
    typeof data.current === 'number',
    'Edit hit dice `current` must be a number'
  );

  invariant(
    typeof data.maximum === 'number',
    'Edit hit dice `maximum` must be a number'
  );

  dispatch({
    type: 'HIT_DICE_EDIT',
    data,
  });
};

export const deleteHitDice = data => dispatch => {
  invariant(typeof data.id === 'string', 'Edit hit dice `id` must be a string');

  dispatch({
    type: 'HIT_DICE_DELETE',
    data,
  });
};
