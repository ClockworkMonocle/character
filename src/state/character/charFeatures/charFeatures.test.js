import { createState } from 'state';
import { createFeature, editFeature, deleteFeature } from './actions';

test('Creating/Editing features and class charges must have correct mapping', () => {
  const mockFeature = {
    id: 'feat-1',
    classChargesId: 'class-charges-1',
  };
  const wrongClassChargeFeature = {
    id: 'feat-1',
    classChargesId: 'does-not-exist',
  };
  const mockCharge = {
    id: 'class-charge-1',
    featureId: 'feat-1',
  };
  const incompleteClassCharge = {
    id: 'class-charges-1',
  };
  const wrongClassCharge = {
    id: 'class-charges-1',
    featureId: 'some-other-feature',
  };
  expect(
    createFeature(mockFeature, incompleteClassCharge)
  ).toThrowErrorMatchingSnapshot();
  expect(
    editFeature(mockFeature, incompleteClassCharge)
  ).toThrowErrorMatchingSnapshot();
  expect(
    createFeature(mockFeature, wrongClassCharge)
  ).toThrowErrorMatchingSnapshot();
  expect(
    editFeature(mockFeature, wrongClassCharge)
  ).toThrowErrorMatchingSnapshot();
  expect(
    createFeature(wrongClassChargeFeature, mockCharge)
  ).toThrowErrorMatchingSnapshot();
  expect(
    editFeature(wrongClassChargeFeature, mockCharge)
  ).toThrowErrorMatchingSnapshot();
});

test('Can create a feature without a class charge', () => {
  const store = createState();
  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
  };

  store.dispatch(createFeature(mockFeature));
  const features = store.getState().character.get('charFeatures');
  const classCharges = store.getState().character.get('charClassCharges');
  expect(features.size).toEqual(1);
  expect(features.get(0).toJS()).toEqual(mockFeature);
  expect(classCharges.size).toEqual(0);
});

test('Creating feature with a class charge updates both attributes', () => {
  const store = createState();

  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
    classChargesId: 'charge-1',
  };

  const mockClassCharge = {
    id: 'charge-1',
    name: 'test charges',
    charges: 10,
    current: 10,
    featureId: 'feat-1',
  };

  store.dispatch(createFeature(mockFeature, mockClassCharge));

  let afterFeats = store.getState().character.get('charFeatures');
  let afterChrgs = store.getState().character.get('charClassCharges');

  expect(afterFeats.size).toEqual(1);
  expect(afterChrgs.size).toEqual(1);
  expect(afterChrgs.getIn([0, 'id'])).toEqual('charge-1');
  expect(afterChrgs.getIn([0, 'featureId'])).toEqual('feat-1');
  expect(afterFeats.getIn([0, 'classChargesId'])).toEqual('charge-1');
});

test('Can delete a feature', () => {
  const store = createState();

  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
  };

  store.dispatch(createFeature(mockFeature));
  store.dispatch(deleteFeature({ id: 'feat-1' }));
  const features = store.getState().character.get('charFeatures');
  expect(features.size).toEqual(0);
});

test('Deleting a feature deletes the corresponding class charge', () => {
  const store = createState();

  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
    classChargesId: 'charge-1',
  };

  const mockClassCharge = {
    id: 'charge-1',
    name: 'test charges',
    charges: 10,
    current: 10,
    featureId: 'feat-1',
  };

  store.dispatch(createFeature(mockFeature, mockClassCharge));
  store.dispatch(deleteFeature({ id: 'feat-1', classChargesId: 'charge-1' }));
  const charges = store.getState().character.get('charClassCharges');
  const features = store.getState().character.get('charFeatures');
  expect(charges.size).toEqual(0);
  expect(features.size).toEqual(0);
});

test('Can edit a feature with new name and description', () => {
  const store = createState();
  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
  };
  const mockUpdatedFeature = {
    ...mockFeature,
    name: 'new name!',
    desc: 'Another test for editing',
    type: 'PASSIVE',
  };
  store.dispatch(createFeature(mockFeature));
  store.dispatch(editFeature(mockUpdatedFeature));
  const features = store.getState().character.get('charFeatures');
  expect(features.size).toEqual(1);
  expect(features.get(0).toJS()).toEqual(mockUpdatedFeature);
});

test('Editing a feature that does not exist does nothing', () => {
  const store = createState();
  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
  };
  const mockUpdatedFeature = {
    id: 'feat-2',
    name: 'new name!',
    desc: 'Another test for editing',
    type: 'PASSIVE',
  };
  store.dispatch(createFeature(mockFeature));
  store.dispatch(editFeature(mockUpdatedFeature));
  const features = store.getState().character.get('charFeatures');
  expect(features.size).toEqual(1);
  expect(features.get(0).toJS()).toEqual(mockFeature);
});

test('Editng a feature init without CC will have a class charge', () => {
  const store = createState();
  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
  };
  const mockUpdatedFeature = {
    id: 'feat-1',
    name: 'new name!',
    desc: 'Another test for editing',
    type: 'ATTACK',
    classChargesId: 'charge-1',
  };
  const mockClassCharge = {
    id: 'charge-1',
    name: 'test charges',
    charges: 10,
    current: 10,
    featureId: 'feat-1',
  };

  store.dispatch(createFeature(mockFeature));
  store.dispatch(editFeature(mockUpdatedFeature, mockClassCharge));
  const features = store.getState().character.get('charFeatures');
  const charges = store.getState().character.get('charClassCharges');

  expect(features.size).toEqual(1);
  expect(features.get(0).toJS()).toEqual(mockUpdatedFeature);
  expect(charges.size).toEqual(1);
  expect(charges.get(0).toJS()).toEqual(mockClassCharge);
});

test('Editng a CC feature will update the feature and class charge', () => {
  const store = createState();
  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
    classChargesId: 'charge-1',
  };
  const mockClassCharge = {
    id: 'charge-1',
    name: 'test charges',
    charges: 10,
    current: 10,
    featureId: 'feat-1',
  };
  const mockUpdatedFeature = {
    ...mockFeature,
    type: 'PASSIVE',
  };
  const mockUpdatedClassCharge = {
    ...mockClassCharge,
    name: 'other class charge test name',
  };

  store.dispatch(createFeature(mockFeature, mockClassCharge));
  store.dispatch(editFeature(mockUpdatedFeature, mockUpdatedClassCharge));
  const features = store.getState().character.get('charFeatures');
  const charges = store.getState().character.get('charClassCharges');

  expect(features.size).toEqual(1);
  expect(features.get(0).toJS()).toEqual(mockUpdatedFeature);
  expect(charges.size).toEqual(1);
  expect(charges.get(0).toJS()).toEqual(mockUpdatedClassCharge);
});

test('Editing a CC feature to remove a class charge, removes the charge', () => {
  const store = createState();

  const mockFeature = {
    id: 'feat-1',
    name: 'test feat',
    desc: 'some test feat',
    type: 'ATTACK',
    classChargesId: 'charge-1',
  };
  const mockClassCharge = {
    id: 'charge-1',
    name: 'test charges',
    charges: 10,
    current: 10,
    featureId: 'feat-1',
  };
  const mockUpdatedFeature = {
    ...mockFeature,
    classChargesId: undefined,
  };

  store.dispatch(createFeature(mockFeature, mockClassCharge));
  store.dispatch(editFeature(mockUpdatedFeature));

  const features = store.getState().character.get('charFeatures');
  const charges = store.getState().character.get('charClassCharges');

  expect(charges.size).toEqual(0);
  expect(features.getIn([0, 'classChargesId'])).toEqual(undefined);
  expect(features.get(0).toJS()).toEqual(mockUpdatedFeature);
});
