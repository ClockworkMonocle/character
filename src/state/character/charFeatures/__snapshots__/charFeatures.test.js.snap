// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating/Editing features and class charges must have correct mapping 1`] = `"Creating a feature with class charge requires \`featureId\` to be a string"`;

exports[`Creating/Editing features and class charges must have correct mapping 2`] = `"Creating a feature with class charge requires \`featureId\` to be a string"`;

exports[`Creating/Editing features and class charges must have correct mapping 3`] = `"Class charge \`featureId\` must equal the feauture id"`;

exports[`Creating/Editing features and class charges must have correct mapping 4`] = `"Class charge \`featureId\` must equal the feauture id"`;

exports[`Creating/Editing features and class charges must have correct mapping 5`] = `"A feature's \`classChargesId\` must match the class charge \`id\`"`;

exports[`Creating/Editing features and class charges must have correct mapping 6`] = `"A feature's \`classChargesId\` must match the class charge \`id\`"`;
